File: artifacts/api-server/src/routes/reserve.py
import math
import random
from datetime import datetime, timedelta
from typing import List
from fastapi import APIRouter, Depends, Body, HTTPException
from sqlalchemy.orm import Session
from ..database import get_db
from ..models import ReserveStatus
from ..schemas import ReserveStatusResponse

router = APIRouter()

@router.get("/reserve/status", response_model=ReserveStatusResponse)
def get_reserve_status(db: Session = Depends(get_db)):
    status = db.query(ReserveStatus).first()
    if not status:
        raise HTTPException(status_code=404, detail="Reserve status not found")
    return status

@router.post("/reserve/optimize")
def optimize_reserve(
    db: Session = Depends(get_db),
    scenario_id: str = Body("hormuz_closure", alias="scenarioId"),
    drawdown_rate_mbpd: float = Body(0.3, alias="drawdownRateMbpd")
):
    current = db.query(ReserveStatus).first()
    current_days = current.current_days if current else 9.5
    daily_consumption = 5.2  # MBPD

    if scenario_id == "hormuz_closure":
        supply_gap_mbpd = 2.1
    elif scenario_id == "opec_cut":
        supply_gap_mbpd = 0.8
    elif scenario_id == "red_sea_suspension":
        supply_gap_mbpd = 1.4
    else:
        supply_gap_mbpd = 0.6

    net_drawdown = max(0.0, drawdown_rate_mbpd)
    
    if net_drawdown > 0:
        days_until_depletion = math.floor((current_days * daily_consumption) / net_drawdown)
    else:
        days_until_depletion = None

    today = datetime.now()
    
    if days_until_depletion is not None:
        depletion_date = (today + timedelta(days=days_until_depletion)).strftime("%Y-%m-%d")
    else:
        depletion_date = None

    replenish_start = (today + timedelta(days=45)).strftime("%Y-%m-%d")
    recommended_rate = min(supply_gap_mbpd * 0.85, 0.6)

    if net_drawdown > 0.5 or scenario_id == "hormuz_closure":
        depletion_str = str(days_until_depletion) if days_until_depletion is not None else "N/A"
        action_banner = (
            f"CRITICAL ACTION: Authorize SPR drawdown at {recommended_rate:.2f} MBPD. "
            f"Notify Mangalore, Padur, and Visakhapatnam SPR facilities. "
            f"Estimated depletion: {depletion_str} days at current rate. "
            f"Begin emergency replenishment procurement immediately."
        )
    elif supply_gap_mbpd > 0.5:
        action_banner = (
            f"ELEVATED ALERT: Pre-position SPR drawdown authorization. "
            f"Recommended rate: {recommended_rate:.2f} MBPD. "
            f"Monitor supply gap — activate within 72 hours if situation deteriorates."
        )
    else:
        action_banner = (
            f"MONITOR: No immediate SPR drawdown required. Maintain current reserve levels. "
            f"Schedule routine replenishment window beginning {replenish_start}."
        )

    return {
        "recommendedDrawdownRate": recommended_rate,
        "depletionDate": depletion_date,
        "repletionDate": None,
        "replenishmentStart": replenish_start,
        "actionBanner": action_banner,
    }

@router.get("/reserve/drawdown-schedule")
def get_drawdown_schedule(db: Session = Depends(get_db)):
    current = db.query(ReserveStatus).first()
    start_days = current.current_days if current else 9.5
    daily_demand = 5.2

    schedule = []
    today = datetime.now()
    reserve_level = start_days * daily_demand

    for i in range(30):
        date_str = (today + timedelta(days=i)).strftime("%Y-%m-%d")
        demand_forecast = daily_demand + (random.random() * 0.4 - 0.2)
        drawdown = 0.35 if i < 15 else 0.0
        supply_gap = 1.8 - i * 0.12 if i < 10 else 0.0
        
        reserve_level = max(0.0, reserve_level - drawdown)

        schedule.append({
            "date": date_str,
            "reserveLevel": round(reserve_level, 2),
            "demandForecast": round(demand_forecast, 2),
            "supplyGap": round(supply_gap, 2),
        })

    return schedule

File: artifacts/api-server/src/routes/reserve.ts
import { Router, type IRouter } from "express";
import { db, reserveStatusTable } from "@workspace/db";

const router: IRouter = Router();

router.get("/reserve/status", async (_req, res): Promise<void> => {
  const [status] = await db.select().from(reserveStatusTable).limit(1);
  if (!status) {
    res.status(404).json({ error: "Reserve status not found" });
    return;
  }
  res.json({
    currentDays: status.currentDays,
    capacityDays: status.capacityDays,
    fillPercent: status.fillPercent,
    drawdownRate: status.drawdownRate,
    recommendedAction: status.recommendedAction,
    riskOfDepletion: status.riskOfDepletion,
    replenishmentWindow: status.replenishmentWindow,
  });
});

router.post("/reserve/optimize", async (req, res): Promise<void> => {
  const { scenarioId = "hormuz_closure", drawdownRateMbpd = 0.3 } = req.body ?? {};

  const [current] = await db.select().from(reserveStatusTable).limit(1);
  const currentDays = current?.currentDays ?? 9.5;
  const dailyConsumption = 5.2; // MBPD

  const supplyGapMbpd = scenarioId === "hormuz_closure" ? 2.1 :
    scenarioId === "opec_cut" ? 0.8 :
    scenarioId === "red_sea_suspension" ? 1.4 : 0.6;

  const netDrawdown = Math.max(0, drawdownRateMbpd);
  const daysUntilDepletion = netDrawdown > 0
    ? Math.floor((currentDays * dailyConsumption) / netDrawdown)
    : null;

  const depletionDate = daysUntilDepletion
    ? new Date(Date.now() + daysUntilDepletion * 86400000).toISOString().split("T")[0]
    : null;

  const replenishStart = new Date(Date.now() + 45 * 86400000).toISOString().split("T")[0];

  const recommendedRate = Math.min(supplyGapMbpd * 0.85, 0.6);

  let actionBanner: string;
  if (netDrawdown > 0.5 || (scenarioId === "hormuz_closure")) {
    actionBanner = `CRITICAL ACTION: Authorize SPR drawdown at ${recommendedRate.toFixed(2)} MBPD. Notify Mangalore, Padur, and Visakhapatnam SPR facilities. Estimated depletion: ${daysUntilDepletion ?? 'N/A'} days at current rate. Begin emergency replenishment procurement immediately.`;
  } else if (supplyGapMbpd > 0.5) {
    actionBanner = `ELEVATED ALERT: Pre-position SPR drawdown authorization. Recommended rate: ${recommendedRate.toFixed(2)} MBPD. Monitor supply gap — activate within 72 hours if situation deteriorates.`;
  } else {
    actionBanner = `MONITOR: No immediate SPR drawdown required. Maintain current reserve levels. Schedule routine replenishment window beginning ${replenishStart}.`;
  }

  res.json({
    recommendedDrawdownRate: recommendedRate,
    depletionDate,
    repletionDate: null,
    replenishmentStart: replenishStart,
    actionBanner,
  });
});

router.get("/reserve/drawdown-schedule", async (_req, res): Promise<void> => {
  const [current] = await db.select().from(reserveStatusTable).limit(1);
  const startDays = current?.currentDays ?? 9.5;
  const dailyDemand = 5.2;

  const schedule = [];
  const today = new Date();
  let reserveLevel = startDays * dailyDemand;

  for (let i = 0; i < 30; i++) {
    const date = new Date(today);
    date.setDate(today.getDate() + i);
    const demandForecast = dailyDemand + (Math.random() * 0.4 - 0.2);
    const drawdown = i < 15 ? 0.35 : 0;
    const supplyGap = i < 10 ? 1.8 - i * 0.12 : 0;
    reserveLevel = Math.max(0, reserveLevel - drawdown);

    schedule.push({
      date: date.toISOString().split("T")[0],
      reserveLevel: parseFloat(reserveLevel.toFixed(2)),
      demandForecast: parseFloat(demandForecast.toFixed(2)),
      supplyGap: parseFloat(supplyGap.toFixed(2)),
    });
  }

  res.json(schedule);
});

export default router;
